### ISSUE TITLE
Join types are somewhat non-intuitive

### ISSUE THREAD

**Issue by lorefnon** [juanluispaz/ts-sql-query#45]
Apologies for the vague issue title, but I have found that when constructing joins dynamically, the types are somewhat hard to work with in practice.

I'll attempt to illustrate with examples using the tables defined in [this file](https://github.com/juanluispaz/ts-sql-query/blob/master/src/examples/PgExample.ts):  

Let's say I want to select customers, and want to join with company to filter the result set if a companyName is available from request: 

```ts
        let customersQ = connection.selectFrom(tCustomer);

        let companyName: string | undefined; // = req.query.companyName
    
        if (companyName) {
            customersQ = customersQ.join(tCompany)
                .on(tCompany.id.equals(tCustomer.companyId))
                // ^ Type error: 
                // typescript: Property 'from' is missing in type 'DynamicOnExpression<DB<"DBConnection">, TCustomer | TCompany, NoTableOrViewRequiredView<DB<"DBConnection">>, never>' but required in type 'SelectExpression<DB<"DBConnection">, TCustomer, NoTableOrViewRequiredView<DB<"DBConnection">>, never>'.
        }

        const customers = customersQ
            .select(extractColumnsFrom(tCustomer))
            .executeSelectMany()
```

This error defies common intuition because conceptually customersQ is an object representing a selection from table of customers. After joining conceptually I get a another table selection which has union of columns from customers and company, so ideally it should be assignable to the previous type - because in SQL whatever selection I can perform from customers table, I can also perform on the join. 

Attempt 2 to workaround: 

```ts
        let customersQ = connection.selectFrom(tCustomer);

        let companyName: string | undefined; // = req.query.companyName
   
        let customersCompanyQ = companyName
            ? customersQ.join(tCompany).on(tCompany.id.equals(tCustomer.companyId)) 
            : customersQ;

        const customers = customersCompanyQ
            .select(extractColumnsFrom(tCustomer))
            // ^ Type error: 
            // This expression is not callable. 
            // Each member of the union type '(<COLUMNS extends SelectColumns<DB<"DBConnection">, TCompany | TCustomer>>(columns: COLUMNS) => ExecutableSelectExpressionWithoutWhere<DB<"DBConnection">, TCompany | TCustomer, COLUMNS, ResultObjectValues<...>, NoTableOrViewRequiredView<...>, never>) | (<COLUMNS extends SelectColumns<...>>(columns: COLUMNS) => Execu...' has signatures, but none of those signatures are compatible with each other.                                                                     

            .executeSelectMany()

```

So even though I don't care about the columns from the table being joined I cant easily perform a join conditionally, without doing an explicit type cast like: 

```ts
        let customersCompanyQ = companyName
            ? customersQ.join(tCompany).on(tCompany.id.equals(tCustomer.companyId)) as unknown as typeof customersQ
            : customersQ;

```

Do you think it is possible to avoid the need for this ?

**Comment by juanluispaz** [comment]
Hi,

----

I will assume your example query is trying to get the list of customers, but if a company name is provided, you only want the customers that belong to a company with that name.

Then, if I write the two cases, the queries will look like:

- In case of no company name is provided:

```ts
const customers = await connection.selectFrom(tCustomer)
    .select(extractColumnsFrom(tCustomer))
    .executeSelectMany()
```

And the query generated for sqlite will be:

```sql
select id as id, first_name as firstName, last_name as lastName, birthday as birthday, company_id as companyId 
from customer
```

- In case of a company name is provided:

```ts
const customers = await connection.selectFrom(tCustomer)
    .join(tCompany).on(tCompany.id.equals(tCustomer.companyId))
    .where(tCompany.name.equals('companyName'))
    .select(extractColumnsFrom(tCustomer))
    .executeSelectMany()
```

And the query generated for sqlite will be:

```sql
select customer.id as id, customer.first_name as firstName, customer.last_name as lastName, customer.birthday as birthday, customer.company_id as companyId 
from customer join company on company.id = customer.company_id 
where company.name = ?
```

Of course, this example is easy to split into two queries, but that is undesired in complex situations.

----

SQL is a declarative language, and ts-sql-query try to preserve it; if you find yourself writing a query using imperative instructions (like an if), most probably that is the wrong way. Another reason for ts-sql-query to be declarative over imperative is that it is the only way to preserve the type safety in TypeScript in the queries.

ts-sql-query offers a lot of different ways to [dynamic queries](https://ts-sql-query.readthedocs.io/en/stable/queries/dynamic-queries/); using those capabilities, I can rewrite the previous queries into a single one that allows me to write even more complex queries:

```ts
const customers = await connection.selectFrom(tCustomer)
    .optionalJoin(tCompany).on(tCompany.id.equals(tCustomer.companyId))
    .where(tCompany.name.equalsIfValue(companyName))
    .select(extractColumnsFrom(tCustomer))
    .executeSelectMany()
```

In case `companyName` have value, the generated query for sqlite is:

```sql
select customer.id as id, customer.first_name as firstName, customer.last_name as lastName, customer.birthday as birthday, customer.company_id as companyId 
from customer join company on company.id = customer.company_id 
where company.name = ?
```

But, in case `companyName` have no value (the value is null or undefined), the generated query for sqlite is:

```sql
select id as id, first_name as firstName, last_name as lastName, birthday as birthday, company_id as companyId 
from customer
```

The way that this work is:
- A join marked as optional will be included only if the joined table/view is used in the query
- The `equalsIfValue` returns a boolean value, but in case of no value (null or undefined), the whole expression is ignored; in consequence the table `tCompany` is not used in the query (and then omitted)

Be aware, that you can combine `IfValue` functions to build complex expressions, ts-sql-query will discard the parts not required when there is no value, but the complex expression will be respected. Example:

```ts
const customers = await connection.selectFrom(tCustomer)
        .optionalJoin(tCompany).on(tCompany.id.equals(tCustomer.companyId))
        .where(
                tCompany.name.equalsIfValue(companyName)
            .or(tCompany.id.equalsIfValue(companyId)))
        .select(extractColumnsFrom(tCustomer))
        .executeSelectMany()
```

*Note*: Testing the solution I put here, I found an issue with the `IfValue` functions that didn't allow the optional join to be omitted. I published ts-sql-query 1.30.0 fixing it; you will need to update to this version to use what I describe here.

**Comment by lorefnon** [comment]
Thanks for the detailed response. Appreciate the explanation - I somehow missed the optionalJoin method in the docs.
